### PR TITLE
Fix the bug that wrongly use the GCM when GSF is not fully available.

### PR DESCRIPTION
**Pull request by elvishew**
When I am using Parse push notification, I found I always could not receive any push notification if I am using Huawei phones, and finally I found out that the Huawei phone has google service framework, so Parse would choose GCM as push type, but unfortunately, there is no Play Store in the phone, and somehow when Parse try to register the GCM push, it always got GSF_PACKAGE_NOT_AVAILABLE error, so the push notification won't work at all, but if I install Play Store on the phone, the GCM works.
I don't know whether this is a bug of Huawei's OS, or a bug of Parse, I just fix it. If we can not find the GCM register service, we regard the phone as GCM register unavailable.
With this fix, even there is no Play Store installed in Huawei phones, the push can work in a PPNS mode.

### CODE REVIEW

File: Parse/src/main/java/com/parse/ManifestInfo.java [hunk 1]
```diff
@@ -466,6 +468,13 @@ private static boolean isGooglePlayServicesAvailable() {
     return Build.VERSION.SDK_INT >= 8 && getPackageInfo("com.google.android.gsf") != null;
   }
 
+  private static boolean isGcmRegisterServiceAvailable() {
+    Intent intent = new Intent(GcmRegistrar.REGISTER_ACTION);
+    intent.setPackage("com.google.android.gsf");
+    List<ResolveInfo> services = getContext().getPackageManager().queryIntentServices(intent, 0);
+    return services != null && services.size() > 0;
+  }
```
Comment (grantland) [inline Parse/src/main/java/com/parse/ManifestInfo.java]: let's move this method into `GcmRegistrar`

**Comment by grantland** [comment]
With this change, we shouldn't need `isGooglePlayServicesAvailable()`, right?

**Comment by grantland** [comment]
I'm not too familiar with this code, @bnham would you have any insight into this?

File: Parse/src/main/java/com/parse/ManifestInfo.java [hunk 2]
```diff
@@ -462,10 +461,6 @@ private static boolean hasAnyGcmSpecificDeclaration() {
     return false;
   }
 
-  private static boolean isGooglePlayServicesAvailable() {
-    return Build.VERSION.SDK_INT >= 8 && getPackageInfo("com.google.android.gsf") != null;
```
Comment (bnham) [inline Parse/src/main/java/com/parse/ManifestInfo.java]: Hah, this reminds me that this code was originally written to work on Donut phones.

Comment (bnham) [inline Parse/src/main/java/com/parse/ManifestInfo.java]: Can we simplify this PR by deleting these two lines here and replacing it with the code you wrote for `isGcmRegisterServiceAvailable`?

**Comment by bnham** [comment]
This looks good, but let's try to minimize the amount of code changed. Can you change your diff so that it just deletes the body of `isGooglePlayServicesAvailable` and replaces it with the body of `isGcmRegisterServiceAvailable`? Then you wouldn't need to change all of those other lines of code.

**Comment by facebook-github-bot-7** [comment]
@elvishew updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@elvishew updated the pull request.

**Comment by rogerhu** [comment]
Do we still need this PR?

**Comment by Jawnnypoo** [comment]
This PR is very out of date, if OP wants to redo it, feel free. Going to close this one though

**Pull request closed without merging**